Citing Rfam
===========

Rfam makes use of a large amount of publicly available data, especially published multiple sequence alignments and secondary structures, and repackages these data in a single searchable and sustainable resource. We have made every effort to credit individual sources on family pages. If you find any of the data presented here useful, please also be sure to credit the primary source also.

Rfam references
---------------
`Rfam 14: expanded coverage of metagenomic, viral and microRNA families <https://doi.org/10.1093/nar/gkaa1047>`_
	| I. Kalvari, E.P. Nawrocki, N. Ontiveros-Palacios, J. Argasinska, K. Lamkiewicz, M. Marz, S. Griffiths-Jones, C. Toffano-Nioche, 
	| D. Gautheret, Z. Weinberg, E. Rivas, S.R. Eddy, R.D. Finn, A. Bateman, and A.I. Petrov
	| **Nucleic Acids Research** (2020) doi: 10.1093/nar/gkaa1047

`Rfam 13.0: shifting to a genome-centric resource for non-coding RNA families <https://academic.oup.com/nar/article/4588106>`_
	| I. Kalvari, J. Argasinska, N. Quinones-Olvera, E.P. Nawrocki, E. Rivas, S.R. Eddy, A. Bateman, R.D. Finn, and A.I. Petrov
	| **Nucleic Acids Research** (2017) doi: 10.1093/nar/gkx1038

`Non‐coding RNA analysis using the Rfam database <https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6754622/>`_
	| I. Kalvari, E.P. Nawrocki, J. Argasinska, N. Quinones‐Olvera, R.D. Finn, A. Bateman and A.I. Petrov
	| **Current Protocols in Bioinformatics** (2018) e51. doi: 10.1002/cpbi.51

`Rfam 12.0: updates to the RNA families database <http://nar.oxfordjournals.org/content/43/D1/D130>`_
	| E.P. Nawrocki, S.W. Burge, A. Bateman, J. Daub, R.Y. Eberhardt, S.R. Eddy, E.W. Floden, P.P. Gardner, T.A. Jones, J.T. and R.D. Finn
	| **Nucleic Acids Research** (2014) doi: 10.1093/nar/gku1063

`Rfam 11.0: 10 years of RNA families <http://nar.oxfordjournals.org/content/41/D1/D226>`_
	| S.W. Burge, J. Daub, R. Eberhardt, J. Tate, L. Barquist, E.P. Nawrocki, S.R. Eddy, P.P. Gardner and A. Bateman.
	| **Nucleic Acids Research** (2012) doi: 10.1093/nar/gks1005

`Rfam: Wikipedia, clans and the “decimal” release <http://nar.oxfordjournals.org/content/39/suppl_1/D141>`_
	| P.P. Gardner, J. Daub, J. Tate, B.L. Moore, I.H. Osuch, S. Griffiths-Jones, R.D. Finn, E.P. Nawrocki, D.L. Kolbe, S.R. Eddy and A. Bateman.
	| **Nucleic Acids Research** (2011)   doi: 10.1093/nar/gkq1129

`Rfam: updates to the RNA families database <http://nar.oxfordjournals.org/cgi/content/full/gkn766?ijkey=qKsfJKpTARDanrQ&keytype=ref>`_
	| P.P. Gardner, J. Daub, J.G. Tate, E.P. Nawrocki, D.L. Kolbe, S. Lindgreen, A.C. Wilkinson, R.D. Finn, S. Griffiths-Jones, S.R. Eddy and A. Bateman
	| **Nucleic Acids Research** (2008)   Database Issue 37:D136-D140

`The RNA WikiProject: community annotation of RNA families <http://rnajournal.cshlp.org/content/14/12/2462>`_
	| J. Daub, P.P. Gardner, J. Tate, D. Ramsköld, M. Manske, W.G. Scott, Z. Weinberg, S. Griffiths-Jones and A. Bateman
	| **RNA** (2008) 12:2462-2464

`Rfam: annotating non-coding RNAs in complete genomes <http://nar.oxfordjournals.org/cgi/content/full/33/suppl_1/D121>`_
	| S. Griffiths-Jones, S. Moxon, M. Marshall, A. Khanna, S.R. Eddy, A. Bateman
	| **Nucleic Acids Research** (2005) Database Issue 33:D121-D124

`Rfam: an RNA family database <http://nar.oxfordjournals.org/cgi/content/full/31/1/439>`_
	| S. Griffiths-Jones, A. Bateman, M. Marshall, A. Khanna and S.R. Eddy
	| **Nucleic Acids Research** (2003) 31(1):p439-441


Covariance models and stochastic context-free grammars
-------------------------------------------------------
`Annotating functional RNAs in genomes using Infernal <http://www.ncbi.nlm.nih.gov/pubmed/24639160>`_
	| E.P. Nawrocki
	| **Methods in Molecular Biology** (2014) 1097:163-97

`Infernal 1.1: 100-fold Faster RNA Homology Searches <http://www.ncbi.nlm.nih.gov/pubmed/24008419>`_
	| E. P. Nawrocki, S. R. Eddy
	| **Bioinformatics** (2013) 29:2933-2935

`Computational Identification of Functional RNA Homologs in Metagenomic Data <http://www.ncbi.nlm.nih.gov/pubmed/23722291>`_
	| E. P. Nawrocki, S. R. Eddy
	| **RNA Biology** (2013) 10:1170-1179, 2013

`Infernal 1.0: Inference of RNA Alignments <http://www.ncbi.nlm.nih.gov/pubmed/19307242>`_
	| E.P. Nawrocki, D.L. Kolbe, S.R. Eddy
	| **Bioinformatics** (2009) Mar 23. [Epub ahead of print]

`Local RNA Structure Alignment With Incomplete Sequence <http://www.ncbi.nlm.nih.gov/pubmed/19304875>`_
	| D.L. Kolbe, S.R. Eddy
	| **Bioinformatics** (2009) Mar 20. [Epub ahead of print]

`Query-dependent banding (QDB) for faster RNA similarity searches <http://www.ploscompbiol.org/article/info:doi/10.1371/journal.pcbi.0030056>`_
	| E.P. Nawrocki, S.R. Eddy
	| **PLoS Computational Biology**  (2007) 3(3):e56

`A memory-efficient dynamic programming algorithm for optimal alignment of a sequence to an RNA secondary structure <http://www.biomedcentral.com/1471-2105/3/18>`_
	| S.R. Eddy
	| **BMC Bioinformatics**  (2002) 2(3):18

`Biological Sequence Analysis: Probabilistic Models of Proteins and Nucleic Acids <http://eddylab.org/cupbook.html>`_
	| R. Durbin, S.R. Eddy, A. Krogh, G. Mitchison
	| **Cambridge University Press** (1999) ISBN 0-5216-2971-3

`tRNAscan-SE: a program for improved detection of transfer RNA genes in genomic sequence <http://nar.oxfordjournals.org/cgi/content/full/25/5/955>`_
	| T.M. Lowe, S.R. Eddy
	| **Nucleic Acids Research** (1997) 25(5):955-964

`RNA sequence analysis using covariance models <http://nar.oxfordjournals.org/cgi/reprint/22/11/2079>`_
	| S.R. Eddy, R. Durbin
	| **Nucleic Acids Research** (1994) 22(11):2079-88
